sshd-core/src/test/java/org/apache/sshd/client/ProxyTest.java

```diff
@@ -0,0 +1,117 @@
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class ProxyTest extends BaseTestSupport {
+
+    public ProxyTest() {
+        super();
+    }
+
+    @Test
+    public void testProxy() throws Exception {
+        try (SshServer server = setupTestServer();
+             SshServer proxy = setupTestServer();
+             SshClient client = setupTestClient()) {
+
+            server.start();
+            proxy.start();
+            client.start();
+
+            String command = "ls -la";
+            String result;
+            try (ClientSession session = createSession(
+                    client,
+                    "localhost", server.getPort(), "user1", "user1",
+                    "localhost", proxy.getPort(), "user2", "user2")) {
+                ByteArrayOutputStream out = new ByteArrayOutputStream();
+                ByteArrayOutputStream errors = new ByteArrayOutputStream();
+                session.executeRemoteCommand(command, out, errors, StandardCharsets.UTF_8);
+                result = out.toString();
+            }
+            assertEquals(command, result);
+        }
+    }
+
+    @SuppressWarnings("checkstyle:ParameterNumber")
+    protected ClientSession createSession(
+            SshClient client,
+            String host, int port, String user, String password,
+            String proxyHost, int proxyPort, String proxyUser, String proxyPassword)
+            throws java.io.IOException {
+        ClientSession session;
+        if (proxyHost != null) {
+            ClientSession proxySession = client.connect(proxyUser, proxyHost, proxyPort)
+                    .verify(CONNECT_TIMEOUT).getSession();
+            proxySession.addPasswordIdentity(proxyPassword);
+            proxySession.auth().verify(AUTH_TIMEOUT);
+            SshdSocketAddress address = new SshdSocketAddress(host, port);
+            ExplicitPortForwardingTracker tracker = proxySession.createLocalPortForwardingTracker(
+                    SshdSocketAddress.LOCALHOST_ADDRESS, address);
+            SshdSocketAddress bound = tracker.getBoundAddress();
+            session = client.connect(user, bound.getHostName(), bound.getPort())
+                    .verify(CONNECT_TIMEOUT).getSession();
+            session.addCloseFutureListener(f -> IoUtils.closeQuietly(tracker));
+        } else {
+            session = client.connect(user, host, port).verify(CONNECT_TIMEOUT).getSession();
+        }
+        session.addPasswordIdentity(password);
+        session.auth().verify(AUTH_TIMEOUT);
+        return session;
+    }
+
+    @Override
+    protected SshServer setupTestServer() {
+        SshServer sshd = super.setupTestServer();
+        // setup forwarding filter to allow the local port forwarding
+        sshd.setForwardingFilter(AcceptAllForwardingFilter.INSTANCE);
+        // setup an echo command
+        sshd.setCommandFactory((session, command) -> new CommandExecutionHelper(command) {
```

Comment:
Would be better to set up the proxy server with no command factory, or with a factory doing something else. As is, one cannot tell whether the proxy or the server handled the command.

I'm also missing something that verifies that the request _did_ go through the proxy. As it's all on localhost, if some mis-wiring occurs somewhere, it might be possible that the client contacts the server directly...